zerver/lib/test_classes.py

```diff
@@ -1123,6 +1123,32 @@ def assert_length(self, items: Collection[Any], count: int) -> None:
             print(f"\nexpected length: {count}\nactual length: {actual_count}")
             raise AssertionError(f"{str(type(items))} is of unexpected size!")
 
+    @contextmanager
+    def assert_query_count(
+        self, count: int, include_savepoints: bool = False, keep_cache_warm: bool = False
+    ) -> Iterator[None]:
+        """
+        This captures the queries executed and check the total number of queries.
+        Useful when minimizing unnecessary roundtrips to the database is important.
+        """
+        with queries_captured(
+            include_savepoints=include_savepoints, keep_cache_warm=keep_cache_warm
+        ) as queries:
+            yield
+        actual_count = len(queries)
+        if actual_count != count:  # nocoverage
+            print("\nITEMS:\n")
+            for index, query in enumerate(queries):
+                print(f"#{index + 1}\nsql: {str(query['sql'])}\ntime: {query['time']}\n")
+            print(f"expected count: {count}\nactual count: {actual_count}")
+            raise AssertionError(
+                f"""
+    {count} queries expected, got {actual_count}.
+    We audit the number of queries for performance critical code paths.
+    Try avoid adding queries when possible.
```

Comment:
- <details>
  <summary>Try to avoid</summary>
   <img src="https://user-images.githubusercontent.com/56781761/196052570-29088155-a2bd-43f6-b3ff-05c0ee816321.png"> 
   </img>
  </details>